File: src/CircleAI.Memory/InMemoryEpisodicStore.cs
// InMemoryEpisodicStore.cs
//
// Thread-safe in-memory implementation of IEpisodicMemoryStore. Used in
// tests and as the default when no persistent backend (SQLite-vec) is
// configured. All data is lost when the process exits.

using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using CircleAI.Core.Validation;

namespace CircleAI.Memory
{
    /// <summary>
    /// In-memory <see cref="IEpisodicMemoryStore"/>. Thread-safe via a
    /// <c>ReaderWriterLockSlim</c>. Maximum capacity is capped to prevent
    /// unbounded growth on long-running processes.
    /// </summary>
    /// <remarks>
    /// Marked <see cref="VerificationLevel.Reference"/> and gated behind
    /// <c>CIRCLEAI_MEM_CAP_001</c>: the 1000-entry FIFO cap is a default
    /// chosen for tests. Production deployments MUST configure
    /// <c>maxEntries</c> explicitly based on observed memory pressure and
    /// expected retention horizon, or substitute a persistent backend.
    /// </remarks>
    [Experimental("CIRCLEAI_MEM_CAP_001", UrlFormat = "https://github.com/bhengubv/CircleAI/blob/master/docs/experimental.md#{0}")]
    [CircleAIVerificationStatus(VerificationLevel.Reference)]
    public sealed class InMemoryEpisodicStore : IEpisodicMemoryStore
    {
        private readonly int _maxEntries;
        private readonly List<EpisodicMemoryEntry> _entries = new();
        private readonly ReaderWriterLockSlim _lock = new();

        /// <param name="maxEntries">
        /// Cap on stored entries. When exceeded the oldest entries are
        /// evicted (FIFO). Default 1000.
        /// </param>
        public InMemoryEpisodicStore(int maxEntries = 1000)
        {
            if (maxEntries <= 0) throw new ArgumentOutOfRangeException(nameof(maxEntries));
            _maxEntries = maxEntries;
        }

        /// <inheritdoc />
        public Task AddAsync(EpisodicMemoryEntry entry, CancellationToken ct = default)
        {
            ArgumentNullException.ThrowIfNull(entry);
            ct.ThrowIfCancellationRequested();

            _lock.EnterWriteLock();
            try
            {
                _entries.Add(entry);
                // Evict oldest when over capacity.
                while (_entries.Count > _maxEntries)
                    _entries.RemoveAt(0);
            }
            finally { _lock.ExitWriteLock(); }

            return Task.CompletedTask;
        }

        /// <inheritdoc />
        public Task<IReadOnlyList<EpisodicMemoryEntry>> SearchAsync(
            float[]? queryEmbedding,
            int topK = 5,
            CancellationToken ct = default)
        {
            ct.ThrowIfCancellationRequested();

            _lock.EnterReadLock();
            List<EpisodicMemoryEntry> snapshot;
            try { snapshot = new List<EpisodicMemoryEntry>(_entries); }
            finally { _lock.ExitReadLock(); }

            IReadOnlyList<EpisodicMemoryEntry> result;

            if (queryEmbedding is null || queryEmbedding.Length == 0)
            {
                // No embedding — return most recent.
                result = snapshot
                    .OrderByDescending(e => e.RecordedAtUtc)
                    .Take(topK)
                    .ToList();
            }
            else
            {
                // Cosine similarity search (only against entries that have embeddings
                // with the same dimension as the query).
                result = snapshot
                    .Where(e => e.Embedding is not null &&
                                e.Embedding.Length == queryEmbedding.Length)
                    .Select(e => (Entry: e, Score: CosineSimilarity(queryEmbedding, e.Embedding!)))
                    .OrderByDescending(x => x.Score)
                    .Take(topK)
                    .Select(x => x.Entry)
                    .ToList();
            }

            return Task.FromResult(result);
        }

        /// <inheritdoc />
        public Task<IReadOnlyList<EpisodicMemoryEntry>> GetRecentAsync(
            int count = 10,
            CancellationToken ct = default)
        {
            ct.ThrowIfCancellationRequested();

            _lock.EnterReadLock();
            List<EpisodicMemoryEntry> snapshot;
            try { snapshot = new List<EpisodicMemoryEntry>(_entries); }
            finally { _lock.ExitReadLock(); }

            IReadOnlyList<EpisodicMemoryEntry> result = snapshot
                .OrderByDescending(e => e.RecordedAtUtc)
                .Take(count)
                .ToList();

            return Task.FromResult(result);
        }

        /// <inheritdoc />
        public Task<int> CountAsync(CancellationToken ct = default)
        {
            ct.ThrowIfCancellationRequested();
            _lock.EnterReadLock();
            try { return Task.FromResult(_entries.Count); }
            finally { _lock.ExitReadLock(); }
        }

        /// <inheritdoc />
        public Task<int> PruneOlderThanAsync(DateTimeOffset cutoff, CancellationToken ct = default)
        {
            ct.ThrowIfCancellationRequested();

            _lock.EnterWriteLock();
            try
            {
                int before = _entries.Count;
                _entries.RemoveAll(e => e.RecordedAtUtc < cutoff);
                return Task.FromResult(before - _entries.Count);
            }
            finally { _lock.ExitWriteLock(); }
        }

        // ------------------------------------------------------------------
        // Helpers
        // ------------------------------------------------------------------

        private static float CosineSimilarity(float[] a, float[] b)
        {
            // Both vectors are already L2-normalised (written by TextEmbedder),
            // so cosine similarity == dot product.
            float dot = 0f;
            for (int i = 0; i < a.Length; i++) dot += a[i] * b[i];
            return dot;
        }
    }
}

File: python/src/circle_ai/memory/in_memory_episodic_store.py
# memory/in_memory_episodic_store.py
#
# Concrete in-memory IEpisodicMemoryStore for the memory-brain. Ported from
# CircleAI.Memory (InMemoryEpisodicStore) — the C# reference — and mirrors the
# TypeScript pilot (memory/stores.ts) and the Go port (memory_stores.go).
#
# All data is lost when the process exits; a persistent (SQLite-vec) backend is a
# later slice. The algorithms — cosine similarity == dot product on L2-normalised
# vectors, recency fallback, FIFO capacity eviction, prune — are identical to the
# reference.
#
# Async posture: the tree marks IEpisodicMemoryStore's methods ``*_async``
# (asyncio_mode = "auto"). The work here is pure in-memory CPU, so each coroutine
# completes synchronously — no lock is needed (asyncio is single-threaded and the
# body never awaits mid-operation), which is the same reasoning the TS pilot uses
# for dropping the C# ReaderWriterLockSlim.

from __future__ import annotations

from datetime import datetime
from typing import Optional

from .episodic_memory import EpisodicMemoryEntry


class InMemoryEpisodicStore:
    """In-memory :class:`IEpisodicMemoryStore`.

    Capacity is capped (FIFO eviction) to prevent unbounded growth on
    long-running processes. Satisfies the ``IEpisodicMemoryStore`` Protocol in
    :mod:`circle_ai.memory.stores` structurally.
    """

    def __init__(self, max_entries: int = 1000) -> None:
        """
        :param max_entries: Cap on stored entries; when exceeded the oldest are
            evicted (FIFO). Default 1000. Must be positive.
        """
        if max_entries <= 0:
            raise ValueError("max_entries must be positive")
        self._max_entries = max_entries
        self._entries: list[EpisodicMemoryEntry] = []

    async def add_async(
        self, entry: EpisodicMemoryEntry, *, ct: Optional[object] = None
    ) -> None:
        """Append a new entry; evict the oldest (FIFO) when over capacity."""
        if entry is None:
            raise ValueError("entry required")
        self._entries.append(entry)
        while len(self._entries) > self._max_entries:
            self._entries.pop(0)

    async def search_async(
        self,
        query_embedding: Optional[list[float]],
        top_k: int = 5,
        *,
        ct: Optional[object] = None,
    ) -> list[EpisodicMemoryEntry]:
        """Return top-k entries by cosine similarity; fall back to recency.

        With no embedding (``None`` or empty) returns the most recent entries.
        Otherwise ranks by cosine similarity — but only against entries whose
        embedding has the same dimension as the query. Both vectors are assumed
        L2-normalised, so cosine similarity == dot product.
        """
        snapshot = list(self._entries)

        if query_embedding is None or len(query_embedding) == 0:
            # No embedding — return most recent.
            return sorted(
                snapshot, key=lambda e: e.recorded_at_utc, reverse=True
            )[:top_k]

        dim = len(query_embedding)
        scored = [
            (e, _cosine_similarity(query_embedding, e.embedding))
            for e in snapshot
            if e.embedding is not None and len(e.embedding) == dim
        ]
        scored.sort(key=lambda x: x[1], reverse=True)
        return [e for e, _ in scored[:top_k]]

    async def get_recent_async(
        self, count: int = 10, *, ct: Optional[object] = None
    ) -> list[EpisodicMemoryEntry]:
        """Return the *count* most recent entries, newest-first."""
        return sorted(
            self._entries, key=lambda e: e.recorded_at_utc, reverse=True
        )[:count]

    async def count_async(self, *, ct: Optional[object] = None) -> int:
        """Total number of entries currently stored."""
        return len(self._entries)

    async def prune_older_than_async(
        self, cutoff: datetime, *, ct: Optional[object] = None
    ) -> int:
        """Remove all entries older than *cutoff*; return the count removed."""
        before = len(self._entries)
        self._entries = [e for e in self._entries if e.recorded_at_utc >= cutoff]
        return before - len(self._entries)


def _cosine_similarity(a: list[float], b: list[float]) -> float:
    """Cosine similarity of two equal-length, L2-normalised vectors (== dot)."""
    dot = 0.0
    for i in range(len(a)):
        dot += a[i] * b[i]
    return dot
